# Проект по SQL

 __Задача:__
 - Проанализировать базу данных, для формирования ценностного предложения для нового продукта.

__Описание данных:__

__Таблица books__\
Содержит данные о книгах:\
•	book_id — идентификатор книги;\
•	author_id — идентификатор автора;\
•	title — название книги;\
•	num_pages — количество страниц;\
•	publication_date — дата публикации книги;\
•	publisher_id — идентификатор издателя.

__Таблица authors__\
Содержит данные об авторах:\
•	author_id — идентификатор автора;\
•	author — имя автора.

__Таблица publishers__\
Содержит данные об издательствах:\
•	publisher_id — идентификатор издательства;\
•	publisher — название издательства.

__Таблица ratings__\
Содержит данные о пользовательских оценках книг:\
•	rating_id — идентификатор оценки;\
•	book_id — идентификатор книги;\
•	username — имя пользователя, оставившего оценку;\
•	rating — оценка книги.

__Таблица reviews__\
Содержит данные о пользовательских обзорах:\
•	review_id — идентификатор обзора;\
•	book_id — идентификатор книги;\
•	username — имя автора обзора;\
•	text — текст обзора.

In [1]:
# импортируем библиотеки 
import pandas as pd 
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры

db_config = {
    'user':  'praktikum_student', # имя пользователя
    'pwd':  'Sdf4$2;d-d30pp', # пароль
    'host':  'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
    'port': 6432, # порт подключения 
    'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'], 
                                                         db_config['host'], 
                                                         db_config['port'], 
                                                         db_config['db'])

# сохраняем коннектор

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# Задания

## Исследуем таблицы — выведем первые строки

__Таблица books__

In [3]:
query ='''

    SELECT *
    FROM books
'''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

Таблица содержит данные о книгах, имеет 1000 строк и 6 колонок

__Таблица authors__

In [4]:
query ='''

    SELECT *
    FROM authors
'''
pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

Таблица содержит данные об авторах, имеет 636 строк и 2 колонки

__Таблица publishers__

In [5]:
query ='''

    SELECT *
    FROM publishers
'''
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

Таблица содержит данные об издательствах, имеет 340 строк и 2 колонки

__Таблица ratings__

In [6]:
query ='''

    SELECT *
    FROM ratings
'''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

Таблица содержит данные о пользовательских оценках книг, имеет 6456 строк и 4 колонки

__Таблица reviews__

In [7]:
query ='''

    SELECT *
    FROM reviews
'''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

Таблица содержит данные о пользовательских обзорах, имеет 2793 строк и 4 колонки

## Посчитаем, сколько книг вышло после 1 января 2000 года

In [8]:
query = '''

SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'

'''

pd.io.sql.read_sql(query, con = engine)

count
0    819

819 книг вышло после 1 января 2000.

## Посчитаем количество обзоров и среднюю оценку для каждой книги

In [9]:
query = '''

SELECT b.book_id, 
       b.title, 
       COUNT(DISTINCT rv.review_id) AS reviews_count, 
       ROUND(AVG(r.rating),2) AS rating_avg
FROM books AS b

LEFT JOIN reviews AS rv
ON b.book_id = rv.book_id

LEFT JOIN  ratings AS r
ON b.book_id = r.book_id

GROUP BY b.book_id, b.title
ORDER BY reviews_count DESC

'''

pd.io.sql.read_sql(query, con = engine)

book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        963                                Water for Elephants   
2        734                                   The Glass Castle   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...   
4        695  The Curious Incident of the Dog in the Night-Time   
..       ...                                                ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
996      808                            The Natural Way to Draw   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites   
998      221                          Essential Tales and Poems   
999      191  Disney's Beauty and the Beast (A Little Golden...   

     reviews_count  rating_avg  
0                7        3.66  
1                6        3.98  
2                6        4.21  
3                6        4.41  
4                6        4.08  
..             ...         ...  
995              0        3.67  
996              0        3.00  
997              0        5.00  
998              0        4.00  
999              0        4.00  

[1000 rows x 4 columns]

Из полученных данных можно сделать вывод, книги имеют от 7 пользовательских обзоров до 0, и как правило это никак не влияет на среднюю пользовательскую оценку. Например книга Twilight (Twilight #1) имеет больше всего обзоров, нл имеет среднюю оценку 3,66, а книга Essential Tales and Poems вообще не имеет обзоров, но меет среднюю оценку 4.

## Определим издательство, которое выпустило наибольшее число книг толще 50 страниц.

In [10]:
query = '''

SELECT
  publisher,
  COUNT (book_id) as book_count
FROM
  books AS b
LEFT JOIN
  publishers AS p ON p.publisher_id=b.publisher_id
WHERE
  num_pages>50
GROUP BY
  publisher
ORDER BY
  book_count DESC
LIMIT 1

'''

pd.io.sql.read_sql(query, con = engine)

publisher  book_count
0  Penguin Books          42

Согласно полученным данным, издательство которое выпустило наибольшее число книг толще 50 страниц - Penguin Books (42 книги)

## Определим автора с самой высокой средней оценкой книг (только книги с 50 и более оценками)

In [11]:
query = '''

WITH 

tab AS (
SELECT book_id
FROM ratings
GROUP BY book_id
HAVING COUNT(rating_id) > 50)


SELECT a.author_id, 
       a.author, 
       ROUND(AVG(r.rating), 2) AS rating_avg
FROM books AS b

INNER JOIN tab
ON b.book_id = tab.book_id

INNER JOIN authors AS a
ON b.author_id = a.author_id

INNER JOIN ratings AS r
ON tab.book_id = r.book_id

GROUP BY a.author_id, a.author
ORDER BY AVG(r.rating) DESC

LIMIT 1

'''

pd.io.sql.read_sql(query, con = engine)

author_id                      author  rating_avg
0        236  J.K. Rowling/Mary GrandPré        4.29

J.K. Rowling - автор с самой высокой средней оценкой книг.

## Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [12]:
query = '''

SELECT COUNT(DISTINCT rv.text)/ COUNT(DISTINCT rv.username) AS count
FROM ratings AS r

INNER JOIN reviews AS rv
ON r.username = rv.username

WHERE r.username in ( 
                    SELECT username
                    FROM ratings
                    GROUP BY username
                    HAVING COUNT(rating_id) > 50
                    )

'''

pd.io.sql.read_sql(query, con = engine)

count
0     24

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.

# Выводы:

В представленной базе данных содержатся данные о 1000 книг, 819 книг было выпущено после 1 января 2000 года. Книги имеют от 7 пользовательских обзоров до 0, и как правило это никак не влияет на среднюю пользовательскую оценку. Например книга Twilight (Twilight #1) имеет больше всего обзоров, но имеет среднюю оценку 3,66, а книга Essential Tales and Poems вообще не имеет обзоров, но меет среднюю оценку 4. Penguin Books - издательство которое выпустило наибольшее число книг толще 50 страниц. J.K. Rowling - автор с самой высокой средней оценкой. Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.